url4c41254592_topskychen_AuthTopk_tgz-pJ8-quickhull3d_QuickHull3DTestJ8/src/spatialindex/rtree/RTree.java

```diff
@@ -0,0 +1,1674 @@
+
+		m_rootID = ds.readInt();
+		m_treeVariant = ds.readInt();
+		m_fillFactor = ds.readDouble();
+		m_indexCapacity = ds.readInt();
+		m_leafCapacity = ds.readInt();
+		m_nearMinimumOverlapFactor = ds.readInt();
+		m_splitDistributionFactor = ds.readDouble();
+		m_reinsertFactor = ds.readDouble();
+		m_dimension = ds.readInt();
+		m_stats.m_nodes = ds.readLong();
+		m_stats.m_data = ds.readLong();
+		m_stats.m_treeHeight = ds.readInt();
+		destFileNamePoint = DataIO.readString(ds);
+		destFileNameLine = DataIO.readString(ds);
+		
+		for (int cLevel = 0; cLevel < m_stats.m_treeHeight; cLevel++) {
+			m_stats.m_nodesInLevel.add(new Integer(ds.readInt()));
+		}
+	}
+
+	/**
+	 * create overflowTable then insert data
+	 * 
+	 * @param pData
+	 * @param mbr
+	 * @param id
+	 */
+	protected void insertData_impl(byte[] pData, Region mbr, int id, ArrayList<Integer> changedNodeID) {
+		boolean[] overflowTable;
+
+		Stack<Integer> pathBuffer = new Stack<Integer>();
+
+		Node root = readNode(m_rootID);
+
+		overflowTable = new boolean[root.m_level];
+		for (int cLevel = 0; cLevel < root.m_level; cLevel++) {
+			overflowTable[cLevel] = false;
+		}
+
+		Node l = root.chooseSubtree(mbr, 0, pathBuffer);
+		if (changedNodeID != null)
+			for (Integer i : pathBuffer)
+				changedNodeID.add(i);
+		l.insertData(pData, mbr, id, pathBuffer, overflowTable, changedNodeID);
+
+		m_stats.m_data++;
+	}
+
+	/**
+	 * insert data with given overflowTable
+	 * 
+	 * @param pData
+	 * @param mbr
+	 * @param id
+	 * @param level
+	 * @param overflowTable
+	 */
+	protected void insertData_impl(byte[] pData, Region mbr, int id, int level, boolean[] overflowTable, ArrayList<Integer> changedNodeID) {
+		assert mbr.getDimension() == m_dimension;
+
+		Stack<Integer> pathBuffer = new Stack<Integer>();
+
+		Node root = readNode(m_rootID);
+		Node n = root.chooseSubtree(mbr, level, pathBuffer);
+		if (changedNodeID != null)
+			for (Integer i : pathBuffer)
+				changedNodeID.add(i);
+		n.insertData(pData, mbr, id, pathBuffer, overflowTable, changedNodeID);
+	}
+
+	/**
+	 * delete a data
+	 * 
+	 * @param mbr
+	 *            contains the data
+	 * @param id
+	 *            indicates the data
+	 * @return
+	 */
+	protected boolean deleteData_impl(final Region mbr, int id, ArrayList<Integer> changedNodeID) {
+		assert mbr.getDimension() == m_dimension;
+
+		boolean bRet = false;
+
+		Stack<Integer> pathBuffer = new Stack<Integer>();
+
+		Node root = readNode(m_rootID);
+		Leaf l = root.findLeaf(mbr, id, pathBuffer);
+
+		if (l != null) {
+			l.deleteData(id, pathBuffer, changedNodeID);
+			m_stats.m_data--;
+			bRet = true;
+		}
+
+		return bRet;
+	}
+
+	/**
+	 * write the node into file
+	 * 
+	 * @param n
+	 * @return
+	 * @throws IllegalStateException
+	 */
+	protected int writeNode(Node n) throws IllegalStateException {
+		byte[] buffer = null;
+
+		try {
+			//store node into byte
+			buffer = n.store();
+		} catch (IOException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)
Similar issue at line numbers 1416.